cmd/csaf_validator/main.go

```diff
@@ -0,0 +1,130 @@
+// This file is Free Software under the MIT License
+// without warranty, see README.md and LICENSES/MIT.txt for details.
+//
+// SPDX-License-Identifier: MIT
+//
+// SPDX-FileCopyrightText: 2023 German Federal Office for Information Security (BSI) <https://www.bsi.bund.de>
+// Software-Engineering: 2023 Intevation GmbH <https://intevation.de>
+
+// Package main implements the csaf_validator tool.
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/csaf-poc/csaf_distribution/csaf"
+	"github.com/csaf-poc/csaf_distribution/util"
+	"github.com/jessevdk/go-flags"
+)
+
+type options struct {
+	Version                bool     `long:"version" description:"Display version of the binary"`
+	RemoteValidator        string   `long:"validator" description:"URL to validate documents remotely" value-name:"URL"`
+	RemoteValidatorCache   string   `long:"validatorcache" description:"FILE to cache remote validations" value-name:"FILE"`
+	RemoteValidatorPresets []string `long:"validatorpreset" description:"One or more presets to validate remotely" default:"mandatory"`
+}
+
+func main() {
+	opts := new(options)
+
+	parser := flags.NewParser(opts, flags.Default)
+	parser.Usage = "[OPTIONS] files..."
+	files, err := parser.Parse()
+	errCheck(err)
+
+	if opts.Version {
+		fmt.Println(util.SemVersion)
+		return
+	}
+
+	if len(files) == 0 {
+		log.Println("No domains given.")
+		return
+	}
+
+	errCheck(run(opts, files))
+}
+
+// run validates the given files.
+func run(opts *options, files []string) error {
+
+	var validator csaf.RemoteValidator
+
+	if opts.RemoteValidator != "" {
+		validatorOptions := csaf.RemoteValidatorOptions{
+			URL:     opts.RemoteValidator,
+			Presets: opts.RemoteValidatorPresets,
+			Cache:   opts.RemoteValidatorCache,
+		}
+		var err error
+		if validator, err = validatorOptions.Open(); err != nil {
+			return fmt.Errorf(
+				"preparing remote validator failed: %w", err)
+		}
+		defer validator.Close()
+	}
+
+	for _, file := range files {
+		doc, err := loadJSONFromFile(file)
+		if err != nil {
+			log.Printf("error: loading %q as JSON failed: %v\n", file, err)
+			continue
+		}
+		// Validate agsinst Schema.
+		validationErrs, err := csaf.ValidateCSAF(doc)
+		if err != nil {
+			log.Printf("error: validating %q against schema failed: %v\n",
+				file, err)
+
+		}
+		if len(validationErrs) > 0 {
+			fmt.Printf("schema validation errors of %q\n", file)
+			for _, vErr := range validationErrs {
+				fmt.Printf("  * %s\n", vErr)
+			}
+		} else {
+			fmt.Printf("%s passes the schema validation.\n", file)
+		}
+		// Validate against remote validator.
+		validate, err := validator.Validate(doc)
+		if err != nil {
+			return fmt.Errorf("remote validation of %q failed: %w",
+				file, err)
+		}
+		var passes string
+		if validate {
+			passes = "passes"
+		} else {
+			passes = "does not pass"
+		}
+		fmt.Printf("%q %s remote validation.\n", file, passes)
```

Comment:
Please add an option to see the results of the remote validator (as one could use that to run optional or informative tests).

Reply:
The [remote validation code](https://github.com/csaf-poc/csaf_distribution/blob/0c398199303a2bb0ddc76a5046933735789ef115/csaf/remotevalidation.go#L97) currently does not handle details of the validation.

To do so this code has to be extended.

Reply:
@s-l-teichmann Thanks for the hint. We'll track that as new issue #324. If you think that it's in the threshold - feel free to do it in this PR.

Reply:
I would prefer to tackle this in another PR. 

Adding this feature would involve changing the internal caching of the remote results which leads to
to the risk to break the library. Breaking the library could endanger
the other uses in the provider, the downloader, etc.